internal/kibana/slo.go

```diff
@@ -0,0 +1,506 @@
+package kibana
+
+import (
+	"context"
+
+	"github.com/elastic/terraform-provider-elasticstack/generated/slo"
+	"github.com/elastic/terraform-provider-elasticstack/internal/clients"
+	"github.com/elastic/terraform-provider-elasticstack/internal/clients/kibana"
+	"github.com/elastic/terraform-provider-elasticstack/internal/models"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+)
+
+func ResourceSlo() *schema.Resource {
+	sloSchema := map[string]*schema.Schema{
+		"id": {
+			Description: "An ID (8 and 36 characters). If omitted, a UUIDv1 will be generated server-side.",
+			Type:        schema.TypeString,
+			Optional:    true,
+			ForceNew:    true,
+			Computed:    true,
+		},
+		"name": {
+			Description: "The name of the SLO.",
+			Type:        schema.TypeString,
+			Required:    true,
+		},
+		"description": {
+			Description: "A description for the SLO.",
+			Type:        schema.TypeString,
+			Required:    true,
+		},
+		"indicator": {
+			Type:     schema.TypeList,
+			Required: true,
+			MinItems: 1,
+			MaxItems: 1,
+			Elem: &schema.Resource{
+				Schema: map[string]*schema.Schema{
+					"type": {
+						Type:         schema.TypeString,
+						Required:     true,
+						ValidateFunc: validation.StringInSlice([]string{"sli.kql.custom", "sli.apm.transactionErrorRate", "sli.apm.transactionDuration", "sli.histogram.custom", "sli.metric.custom"}, false),
+					},
+					"params": {
```

Comment:
`params` can be represented by, e.g. JSON string (similar to `params` in `elasticstack_kibana_alerting_rule` or `config` in `elasticstack_kibana_action_connector` - internal/kibana/connector.go#42). This should allow having different objects (JSON strings) for `params` attribure based on `type`'s value that makes it possible to support `sli.histogram.custom` and `sli.metric.custom` types.

Reply:
Another approach to support these 2 currently unsupported fields is to replace general `indicator` attribute with more specific indicator attributes (e.g. `histogram_custom_indicator` for `sli.histogram.custom`, `metric_custom_indicator` for `sli.metric.custom`, etc.) and add a validation that only one of them is present in configuration (but at least one should be specified).

Reply:
@Kushmaro, what do you think?

Reply:
> params can be represented by, e.g. JSON string 

I thought about this, but don't we lose the validation at that point? This is a pretty complex piece of the API, and it'd be nice to validate that the params per indicator are correct. 

> Another approach to support these 2 currently unsupported fields is to replace general indicator attribute with more specific indicator attributes (e.g. histogram_custom_indicator for sli.histogram.custom, metric_custom_indicator for sli.metric.custom, etc.) and add a validation that only one of them is present in configuration (but at least one should be specified).

Can you elaborate here? I'm not totally following. Perhaps you can provide some pseudocode?

Reply:
Yes, we will lose some convenience of TF with JSON string.

Regarding the second approach. The idea is to specify separate `indicator` field for each unique use case. E.g. we can define the following indicator attribute for `sli.metric.custom` in addition to the `indicator` attribute that you implemented:

```
               "indicator": {
               // definition omitted
               }
		"metric_custom_indicator": {
			Type:         schema.TypeList,
			Required:     true,
			MinItems:     1,
			MaxItems:     1,
			ExactlyOneOf: []string{"indicator", "metric_custom_indicator", "histogram_custom_indicator"},
			Elem: &schema.Resource{
				Schema: map[string]*schema.Schema{
					"params": {
						Type:     schema.TypeList,
						Required: true,
						MinItems: 1,
						MaxItems: 1,
						Elem: &schema.Resource{
							Schema: map[string]*schema.Schema{
								"index": {
									Type:     schema.TypeString,
									Required: true,
								},
								"filter": {
									Type:     schema.TypeString,
									Optional: true,
								},
								"good": {
									Type:     schema.TypeList,
									Required: true,
									MinItems: 1,
									MaxItems: 1,
									Elem: &schema.Resource{
										Schema: map[string]*schema.Schema{
											"metrics": {
												Type:     schema.TypeList,
												Required: true,
												Elem: &schema.Resource{
													Schema: map[string]*schema.Schema{
														"name": {
															Type:     schema.TypeString,
															Required: true,
														},
														"aggregation": {
															Type:     schema.TypeString,
															Required: true,
														},
														"field": {
															Type:     schema.TypeString,
															Required: true,
														},
														"filter": {
															Type:     schema.TypeString,
															Optional: true,
														},
													},
												},
											},
											"equation": {
												Type:     schema.TypeString,
												Required: true,
											},
										},
									},
								},
								"total": {
									Type:     schema.TypeList,
									Required: true,
									// Elem should be defined similar to "good"
								},
								"timestamp_field": {
									Type:     schema.TypeString,
									Required: true,
								},
							},
						},
					},
				},
			},
		},

               "histogram_custom_indicator": {
              // definition omitted
              }
```

The code implies that the schema defines 3 indicator attributes - `indicator` attribute defined as before covers `sli.kql.custom`, `sli.apm.transactionErrorRate` and `sli.apm.transactionDuration` types, `metric_custom_indicator` covers `sli.metric.custom` and `histogram_custom_indicator` attribute (I didn't provide a definition for it) covers `sli.histogram.custom`.

`ExactlyOneOf` ([help](https://developer.hashicorp.com/terraform/plugin/framework/migrating/attributes-blocks/validators-predefined)) field in the schema definition (it should be added to all 3 indicator attributes) should make sure that TF configuration defines exactly only one of the 3 mentioned indicator attributes. However, I didn't use `ExactlyOneOf` before and generally didn't implement such approach previously so additional check is needed.

Also, if we follow the approach, the schema definition would be quite verbose, so we can extract some attribute definitions to helper functions, e.g. like [here](https://github.com/elastic/terraform-provider-ec/blob/master/ec/ecresource/deploymentresource/elasticsearch/v2/schema.go#L134).

Does it make sense?

Reply:
Yep, makes perfect sense. Thanks for the extra context and example! I'll see if I can make this work.